libclamav/crtmgr.c

```diff
@@ -281,12 +310,130 @@ void crtmgr_free(crtmgr *m)
         crtmgr_del(m, m->crts);
 }
 
-static int crtmgr_rsa_verify(cli_crt *x509, fp_int *sig, cli_crt_hashtype hashtype, const uint8_t *refhash)
+static int _padding_check_PKCS1_type_1(uint8_t **to, int *tlen,
+                                       uint8_t *from, unsigned int flen,
+                                       unsigned int num)
 {
-    int keylen = fp_unsigned_bin_size(&x509->n), siglen = fp_unsigned_bin_size(sig);
-    int ret, j, objlen, hashlen;
-    uint8_t d[513];
-    fp_int x;
+    int i, j;
+    unsigned char *p;
+
+    p = from;
+
+    /*
+     * The format is
+     * 00 || 01 || PS || 00 || D
+     * PS - padding string, at least 8 bytes of FF
+     * D  - data.
+     */
+
+    if (num < 11 ) /* RSA_PKCS1_PADDING_SIZE */
+        return -1;
+
+    /* Accept inputs with and without the leading 0-byte. */
+    if (num == flen) {
+        if ((*p++) != 0x00) {
+            ERR_raise(ERR_LIB_RSA, RSA_R_INVALID_PADDING);
+            cli_dbgmsg("%s: Bad padding\n", __func__);
+            return -1;
+        }
+        flen--;
+    }
+
+    if ((num != (flen + 1)) || (*(p++) != 0x01)) {
+        cli_dbgmsg("%s: Bad block type\n", __func__);
+        return -1;
+    }
+
+    /* scan over padding data */
+    j = flen - 1;               /* one for type. */
+    for (i = 0; i < j; i++) {
+        if (*p != 0xff) {       /* should decrypt to 0xff */
+            if (*p == 0) {
+                p++;
+                break;
+            } else {
+                cli_dbgmsg("%s: Bad header\n", __func__);
+                return -1;
+            }
+        }
+        p++;
+    }
+
+    if (i == j) {
+        cli_dbgmsg("%s: Bad header\n", __func__);
+        return -1;
+    }
+
+    if (i < 8) {
+        cli_dbgmsg("%s: Bad padding\n", __func__);
+        return -1;
+    }
+    i++; /* Skip over the '\0' */
+    j -= i;
+    *tlen = j;
+    *to = p;
+
+    return 0;
+}
+
+static int crtmgr_get_recorv_data(BIGNUM *sig, cli_crt *x509,
```

Comment:
As noted in the commit description, _padding_check_PKCS1_type_1 has been stolen from OpenSSL. I could have used the API for it but it is marked as deprecated so you would gain a warning and it could vanish.
It is more obvious now what it is and what it does (at least to me).

I *think* that ideally the whole verification process could be pushed to OpenSSL at which point crtmgr_rsa_verify()  would not exist and whole certificate parsing is handled by OpenSSL itself. So you would need know the OIDs and which hash function to use.